File: gameserver/network/utils/reader.py
class Reader:
    def __init__(self, buffer: bytearray):
        self.buffer = buffer
        self.offset = 0

    def move_cursor(self, offset):
        self.offset += offset

    def read_byte(self, size):
        result = self.buffer[self.offset:self.offset + size]
        self.offset += size
        return result

    def read_int(self, size):
        number = 0
        for i in range(size):
            number = number | self.buffer[self.offset] << (8 * i)
            self.offset += 1
        return number

    def read_int_1(self):
        return self.read_int(1)

    def read_int_2(self):
        return self.read_int(2)

    def read_int_4(self):
        return self.read_int(4)

    def read_int_8(self):
        return self.read_int(8)

    def read_string(self):
        size = self.read_int_2()
        return self.read_byte(size).decode('utf-8')

    def to_hex_string(self):
        return " ".join("{:02x}".format(c) for c in self.buffer)

File: src/network/utils/reader.js
import {intToBytes, bytesToInt, toHexString} from './bytes-utils.js';


class Reader {
    constructor(data) {
        this.data = data;
        this.position = 0;
    }

    readIntFromBytes(bytesNumber = 2) {
        const bytes = this.data.slice(this.position, this.position + bytesNumber);
        this.position += bytesNumber;
        return bytesToInt(bytes, 'little', false);
    }

    readStringFromBytes(stringLength) {
        const bytes = this.data.slice(this.position, this.position + stringLength);
        this.position += stringLength;
        return new TextDecoder().decode(bytes);
    }

    readFloatFromBytes(bytesNumber = 4) {
        const bytes = this.data.slice(this.position, this.position + bytesNumber);
        this.position += bytesNumber;
        const buffer = new ArrayBuffer(bytesNumber);
        new Uint8Array(buffer).set(bytes);
        const view = new DataView(buffer);
        if (bytesNumber === 4) {
            return view.getFloat32(0, true);
        }
        if (bytesNumber === 8) {
            return view.getFloat64(0, true);
        }
        throw new Error('Unsupported number of bytes for float');
    }

    readString() {
        const stringLength = this.readInt2();
        return this.readStringFromBytes(stringLength);
    }


    readInt1() {
        return this.readIntFromBytes(1);
    }
    readInt4() {
        return this.readIntFromBytes(4);
    }
    readInt2() {
        return this.readIntFromBytes(2);
    }
    readInt8() {
        return this.readIntFromBytes(8);
    }
    readInt16() {
        return this.readIntFromBytes(16);
    }
    readInt32() {
        return this.readIntFromBytes(32);
    }


    toHexString()
    {
        return toHexString(this.data);
    }

}

export default Reader;